compiler/rustc_hir_analysis/src/check/check.rs

```diff
@@ -37,54 +36,32 @@ use {rustc_attr_data_structures as attrs, rustc_hir as hir};
 use super::compare_impl_item::check_type_bounds;
 use super::*;
 
+fn add_abi_diag_help<T: EmissionGuarantee>(abi: ExternAbi, diag: &mut Diag<'_, T>) {
+    if let ExternAbi::Cdecl { unwind } = abi {
+        let c_abi = ExternAbi::C { unwind };
+        diag.help(format!("use `extern {c_abi}` instead",));
+    } else if let ExternAbi::Stdcall { unwind } = abi {
+        let c_abi = ExternAbi::C { unwind };
+        let system_abi = ExternAbi::System { unwind };
+        diag.help(format!(
+            "if you need `extern {abi}` on win32 and `extern {c_abi}` everywhere else, \
+                use `extern {system_abi}`"
+        ));
+    }
+}
+
 pub fn check_abi(tcx: TyCtxt<'_>, hir_id: hir::HirId, span: Span, abi: ExternAbi) {
```

Comment:
If I understand this PR correctly, we still do not call this function for trait declarations. This means that UNSUPPORTED_CALLING_CONVENTIONS is not emitted for them, and so moving cdecl and friends to `Invalid` will be another sudden breaking change with no prior lint. Is that right?

How hard would it be to find a place in hir_analysis (or wherever) to call this on trait declarations, too?

Reply:
Yes, it's possible to hunt down the various places in HIR, assuming nothing warps the HIR so that a slightly different arrangement of data structures contains `extern "{abi}"` checks.

But I think I would rather wait until https://github.com/rust-lang/rust/pull/138164 so I can make `UNSUPPORTED_CALLING_CONVENTIONS` also correct-by-construction (...by checking it during construction).

Reply:
should land today

Reply:
Oh wow, that's amazing. :) Yeah this is fine then (the FIXME should still be added though IMO, to leave *some* transactional evidence in case Jubilee has to drop the ball on this).

Reply:
there we go! :tada: